File: 0010-regular-expression-matching/0010-regular-expression-matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int n = s.size(), m = p.size();
        deque<deque<bool>> dp(n + 2, deque<bool>(m + 2, false));
        dp[0][0] = true;
        for(int i = 0; i <= n; i++){
            for (int j = 0; j < m; j++){ // fix the state
                if(i == n && j + 1 < m  && p[j + 1] == '*'){
                    dp[i][j + 2] = dp[i][j + 2] || dp[i][j];
                    continue;
                }
                
                bool match = (p[j] == '.' || s[i] == p[j]);
                if (j + 1 < m && p[j + 1] == '*') {
                    dp[i][j + 2] = dp[i][j + 2] || dp[i][j];
                    dp[i + 1][j] = dp[i + 1][j] || (dp[i][j] && match);
                } else {
                    dp[i + 1][j + 1] = dp[i + 1][j + 1] || (dp[i][j] && match);
                }
            }
        }
        return dp[n][m];
    }
};

File: 0010-regular-expression-matching/0010-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n+1) for _ in range(m+1)]
        dp[0][0] = True

        for i in range(m+1):
            for j in range(1, n+1):
                if j > 1 and p[j-1] == '*':
                    dp[i][j] = dp[i][j-2] or (i > 0 and (p[j-2] == '.' or s[i-1] == p[j-2]) and dp[i-1][j])
                else:
                    dp[i][j] = i > 0 and dp[i-1][j-1] and (p[j-1] == '.' or s[i-1] == p[j-1])

        return dp[m][n]